File: cpp/LinkedList.cpp
#include <iostream>
#include "Node.cpp"

using namespace std;

class LinkedList {
    private:
        Node *head;

    public:
        LinkedList() {
            this->head=nullptr;
        }

        LinkedList(Node *head) {
            this->head=head;
        }

        void setHead(Node *head) { this->head=head; }

        Node *getHead() { return this->head; }

        Node *createElement() {
            Node *newElement=new Node();
            newElement->getMahasiswa()->input();
            return newElement;
        }

        Node *createElement(Student *mhs) {
            Node *newElement=new Node(mhs);
            return newElement;
        }

        void insertFirst(Node *newElement) {
            if(this->head==nullptr) {
                this->head=newElement;
            } else {
                newElement->setNext(this->head);
                this->head=newElement;
            }
        }

        void insertLast(Node *newElement) {
            if(this->head==nullptr) {
                this->head=newElement;
            } else {
                Node *curr=this->head;
                while(curr->getNext() != nullptr) {
                    curr=curr->getNext();
                }
                curr->setNext(newElement);
            }
        }

        void deleteFirst() {
            Node *del=new Node();
            del=this->head;
            this->head=this->head->getNext();
            delete del;
        }

        void deleteLast() {
            Node *del=new Node();
            Node *curr=this->head;
            while(curr->getNext()->getNext() != nullptr) {
                curr=curr->getNext();
            }
            del=curr->getNext();
            curr->setNext(nullptr);
            delete del;
        }

        void printList() {
            Node *curr=this->head;
            while(curr != nullptr) {
                curr->getMahasiswa()->print();
                curr=curr->getNext();
            }
        }

        void searchString(int cari, string data) {
            /*

            cari = 1; Mencari Nama
            cari = 2; Mencari NPM

            */

            bool found=false;
            Node *curr=this->head;
            string dicari="";
            while(curr != nullptr) {
                switch(cari) {
                    case 1 : dicari=curr->getMahasiswa()->getNama();
                        break;
                    case 2 : dicari=curr->getMahasiswa()->getNpm();
                        break;
                }

                if(dicari==data) {
                    found=true;
                    curr->getMahasiswa()->print();
                }

                curr=curr->getNext();
            }

            if(!found) {
                cout << "Data yang anda cari tidak dapat ditemukan" << endl;
            }
        }

        void searchNilai(int cari, float nilai) {
            /*
            
            cari = 1; Mencari nilai penguji 1
            cari = 2; Mencari nilai penguji 2
            cari = 3; Mencari nilai pembimbing
            cari = 4; Mencari nilai sidang

            */

            bool found=false;
            Node *curr=this->head;
            float dicari=0;

            while(curr != nullptr) {
                switch(cari) {
                    case 1 : dicari=curr->getMahasiswa()->getPenguji1().getNilai();
                        break;
                    case 2 : dicari=curr->getMahasiswa()->getPenguji2().getNilai();
                        break;
                    case 3 : dicari=curr->getMahasiswa()->getPembimbing().getNilai();
                        break;
                    case 4 : dicari=curr->getMahasiswa()->getNilaiSidang().getNilai();
                        break;
                }

                if(dicari==nilai) {
                    found=true;
                    curr->getMahasiswa()->print();
                }

                curr=curr->getNext();
            }

            if(!found) {
                cout << "Data yang anda cari tidak dapat ditemukan" << endl;
            }
        }

        void searchHurufMutu(int cari, char karakter) {
            /*

            cari = 1; Mencari huruf mutu

            */
            bool found=false;
            Node *curr=this->head;
            char dicari='A';

            while(curr != nullptr) {
                switch(cari) {
                    case 1 : dicari=curr->getMahasiswa()->getHurufMutu(curr->getMahasiswa()->getNilaiSidang().getNilai());
                        break;
                }

                if(dicari==karakter) {
                    found=true;
                    curr->getMahasiswa()->print();
                }

                curr=curr->getNext();
            }

            if(!found) {
                cout << "Data yang anda cari tidak dapat ditemukan" << endl;
            }
        }
};

File: py/LinkedList.py
from Node import Node

class LinkedList :
    def __init__(self) :
        self.__head=Node()

    def setHead(self,head) : self.__head=head

    def getHead(self) : return self.__head

    def insertFirst(self) :
        newElement=Node()
        newElement.getMahasiswa().input()

        if self.__head is None :
            self.__head=newElement
        else :
            newElement.setNext(self.__head)
            self.__head=newElement

    def insertLast(self) :
        newElement=Node()
        newElement.getMahasiswa().input()

        if self.__head is None :
            self.__head=newElement
        else :
            curr=self.__head
            while curr.getNext() is not None :
                curr=curr.getNext()
            curr.setNext(newElement)

    def deleteFirst(self) :
        delete=Node()
        delete=self.__head
        self.__head=self.__head.getNext()
        del delete

    def deleteLast(self) :
        delete=Node()
        curr=self.__head
        while curr.getNext().getNext() is not None :
            curr=curr.getNext()
        delete=curr.getNext()
        curr.setNext(None)
        del delete
    
    def printList(self) :
        curr=self.__head
        while curr.getNext() is not None :
            curr.getMahasiswa().print()
            curr=curr.getNext()

    def searchString(self,cari,data) :
        found=False
        curr=self.__head
        dicari=""

        while curr.getNext() is not None :
            if(cari == "Nama") :
                dicari=curr.getMahasiswa().getNama()
            elif(cari == "NPM") :
                dicari=curr.getMahasiswa().getNpm()
            
            if(dicari == data) :
                found=True
                curr.getMahasiswa().print()
            
            curr=curr.getNext()

        if(found == False) :
            print("Data tidak ada yang ditemukan")

    def searchNilai(self,cari,nilai) :
        found=False
        curr=self.__head
        dicari=0.0

        while curr.getNext() is not None :
            if(cari == "Penguji 1") :
                dicari=curr.getMahasiswa().getPenguji1().getNilai()
            elif(cari == "Penguji 2") :
                dicari=curr.getMahasiswa().getPenguji2().getNilai()
            elif(cari == "Pembimbing") :
                dicari=curr.getMahasiswa().getPembimbing().getNilai()
            elif(cari == "Nilai Sidang") :
                dicari=curr.getMahasiswa().getNilaiSidang().getNilai()
            
            if(dicari == nilai) :
                found=True
                curr.getMahasiswa().print()

            curr=curr.getNext()

        if(found == False) :
            print("Data tidak ada yang ditemukan")

    def searchChar(self,cari,karakter) :
        found=False
        curr=self.__head
        dicari=''

        while curr.getNext() is not None :
            if(cari == "Huruf Mutu") : 
                dicari=curr.getMahasiswa().getHurufMutu()

            if(dicari == karakter) :
                found=True
                curr.getMahasiswa().print()
            
            curr=curr.getNext()
        
        if(found == False) :
            print("Data tidak ada yang ditemukan")
